src/BreakpointViewer.cpp

```diff
@@ -0,0 +1,595 @@
+				enabled = false;
+				item->setText("");
+				setBreakpointStatus(type, row, false);
+			}
+			if (!enabled) return;
+			break;
+		}
+		case SLOT: {
+			qint8 ps, ss;
+			if (processSlotField(index, item->text(), ps, ss)) {
+				auto sa = ScopedAssign(userMode, false);
+				item->setText(QString("%1/%2")
+					.arg(ps == -1 ? QChar('X') : QChar('0' + ps))
+					.arg(ss == -1 ? QChar('X') : QChar('0' + ss)));
+			} else {
+				auto sa = ScopedAssign(userMode, false);
+				item->setText("X/X");
+			}
+			if (!enabled) return;
+			break;
+		}
+		case SEGMENT: {
+			qint16 segment;
+			if (processSegmentField(index, item->text(), segment)) {
+				auto sa = ScopedAssign(userMode, false);
+				item->setText(QString("%1").arg(segment == -1 ? "X" : QString::number(segment)));
+			} else {
+				auto sa = ScopedAssign(userMode, false);
+				item->setText("X");
+			}
+			if (!enabled) return;
+			break;
+		}
+		case CONDITION: {
+			auto sa = ScopedAssign(userMode, false);
+			item->setText(item->text().simplified());
+			if (!enabled) return;
+			break;
+		}
+	}
+
+	if (enabled) {
+		if (createBp) {
+			createBreakpoint(type, row);
+		} else {
+			replaceBreakpoint(type, row);
+		}
+	} else if (index != -1) {
+		removeBreakpoint(type, row);
+	} else {
+		assert(index != -1);
+	}
+}
+
+void BreakpointViewer::changeBpTableItem(QTableWidgetItem* item)
+{
+	if (!userMode) return;
+	changeTableItem(BREAKPOINT, item);
+}
+
+void BreakpointViewer::changeWpTableItem(QTableWidgetItem* item)
+{
+	if (!userMode) return;
+	changeTableItem(WATCHPOINT, item);
+}
+
+void BreakpointViewer::setBreakpoints(Breakpoints* bps)
+{
+	breakpoints = bps;
+}
+
+void BreakpointViewer::keepUnusedItems()
+{
+	int row = 0;
+
+	while (row < bpTableWidget->rowCount()) {
+		QTableWidgetItem* enabledItem = bpTableWidget->item(row, ENABLED);
+		bool enabled = enabledItem->checkState() == Qt::Checked;
+		if (enabled) {
+			bpTableWidget->removeRow(row);
+		} else {
+			row++;
+		}
+	}
+
+	while (row < wpTableWidget->rowCount()) {
+		QTableWidgetItem* enabledItem = wpTableWidget->item(row, ENABLED);
+		bool enabled = enabledItem->checkState() == Qt::Checked;
+		if (enabled) {
+			wpTableWidget->removeRow(row);
+		} else {
+			row++;
+		}
+	}
+}
+
+void BreakpointViewer::setRunState()
+{
+	runState = true;
+}
+
+void BreakpointViewer::setBreakState()
+{
+	runState = false;
+}
+
+void BreakpointViewer::sync()
+{
+	items.clear();
+
+	auto sa = ScopedAssign(userMode, false);
+	keepUnusedItems();
+
+	int bpRow = bpTableWidget->rowCount();
+	int wpRow = wpTableWidget->rowCount();
+
+	if (breakpoints->breakpointCount() != -1) {
+		for (int index = 0; index < breakpoints->breakpointCount(); ++index) {
+			const Breakpoint& bp = breakpoints->getBreakpoint(index);
+
+			switch (bp.type) {
+			case Breakpoint::BREAKPOINT: {
+				bpRow = createTableRow(BREAKPOINT);
+				setBreakpointStatus(BREAKPOINT, bpRow, true);
+
+				bool currentSlot = breakpoints->isBreakpoint(breakpoints->getAddress(index));
```

Comment:
(Already mentioned in an earlier review).
This could be simplified to `...->isBreakpoint(bp.address)`.